syntax = "proto3";

import "cloud/blockstore/public/api/protos/rdma.proto";

package NCloud.NBlockStore.NProto;

option go_package = "a.yandex-team.ru/cloud/blockstore/config";

////////////////////////////////////////////////////////////////////////////////

enum EWaitMode
{
    WAIT_MODE_POLL = 0;
    WAIT_MODE_BUSY_WAIT = 1;
    WAIT_MODE_ADAPTIVE_WAIT = 2;
}

message TRdmaServer
{
    uint32 Backlog = 1;
    uint32 QueueSize = 2;
    uint32 MaxBufferSize = 3;
    uint64 KeepAliveTimeout = 4;            // in milliseconds
    EWaitMode WaitMode = 5;
    uint32 PollerThreads = 6;
    uint32 MaxInflightBytes = 7;            // per client
    uint64 AdaptiveWaitSleepDelay = 8;      // in microseconds
    uint64 AdaptiveWaitSleepDuration = 9;   // in microseconds
    bool AlignedDataEnabled = 10;
}

message TRdmaClient
{
    uint32 QueueSize = 1;
    uint32 MaxBufferSize = 2;
    uint32 PollerThreads = 3;
    EWaitMode WaitMode = 4;
    uint64 AdaptiveWaitSleepDelay = 5;      // in microseconds
    uint64 AdaptiveWaitSleepDuration = 6;   // in microseconds
    bool AlignedDataEnabled = 7;
}

message TRdmaTarget
{
    TRdmaEndpoint Endpoint = 1;
    TRdmaServer Server = 2;                 // deprecated
    uint32 WorkerThreads = 3;
}

message TRdmaConfig
{
    bool ClientEnabled = 1;
    TRdmaClient Client = 2;

    bool ServerEnabled = 3;
    TRdmaServer Server = 4;

    bool DiskAgentTargetEnabled = 5;
    TRdmaTarget DiskAgentTarget = 6;
}
